c10/core/DispatchKey.cpp

```diff
@@ -176,76 +218,38 @@ const char* toString(DispatchKey t) {
 std::ostream& operator<<(std::ostream& str, DispatchKey rhs) {
   return str << toString(rhs);
 }
+std::ostream& operator<<(std::ostream& str, BackendBit rhs) {
+  return str << toString(rhs);
+}
 
 // for a given backend key, return the associated autograd key.
 // for non-backend keys, return AutogradOther as a default.
-// Note: it's convenient and fast to return a default here rather than (say)
-// returning an optional<DispatchKey>, or throwing. But it makes callers
-// responsible for either a) enforcing the invariant that only backend keys
-// be passed as arguments, or b) interpreting our return value carefully.
-//
-DispatchKey getAutogradKeyFromBackend(DispatchKey t) {
-  switch (t) {
-    case DispatchKey::CPU:
-      return DispatchKey::AutogradCPU;
-    case DispatchKey::XPU:
-      return DispatchKey::AutogradXPU;
-    case DispatchKey::CUDA:
-      return DispatchKey::AutogradCUDA;
-    case DispatchKey::XLA:
-      return DispatchKey::AutogradXLA;
-    case DispatchKey::Lazy:
-      return DispatchKey::AutogradLazy;
-    case DispatchKey::MLC:
-      return DispatchKey::AutogradMLC;
-    case DispatchKey::HPU:
-      return DispatchKey::AutogradHPU;
-    case DispatchKey::NestedTensor:
-      return DispatchKey::AutogradNestedTensor;
-    case DispatchKey::PrivateUse1:
-      return DispatchKey::AutogradPrivateUse1;
-    case DispatchKey::PrivateUse2:
-      return DispatchKey::AutogradPrivateUse2;
-    case DispatchKey::PrivateUse3:
-      return DispatchKey::AutogradPrivateUse3;
-    default:
-      return DispatchKey::AutogradOther;
-  }
+// Or, for non-backends that do override autograd, handle them specially.
+// See Note [Non-Backends that Override Autograd]
+DispatchKey getAutogradKeyFromBackend(DispatchKey k) {
+  // We want this to return an autograd key. We're relying on the fact that
+  // getAutogradRelatedKeySetFromBackend returns an autograd key + ADInplaceOrView,
+  // and autograd has higher precedence.
+  return getAutogradRelatedKeySetFromBackend(toBackendBit(k), DispatchKeySet(k)).highestPriorityTypeId();
```

Comment:
I ended up setting it up this way when I went down a performance rabbithole, since it happens to  be the case that `getAutogradKeyFromBackend` is only used at registration time, but `getAutogradRelatedKeySetFromBackend` is called at runtime (in the TensorImpl constructor).

But I agree it would be cleaner to go the other way around (hopefully by making this function constexpr). I'll give that another try.

Reply:
Sounds good!